### PR TITLE
chore: improve ci

### PR DESCRIPTION
**Pull request by 0xOsiris**
> [!NOTE]
> **Low Risk**
> CI-only changes (pinning and build tooling); no application runtime code, though Rust test/docs jobs now use a specific nightly and mold which could surface toolchain or linker edge cases.
> 
> **Overview**
> Pins third-party GitHub Actions across workflows and composite actions to **full commit SHAs** (Docker, checkout, cache, AWS, `taiki-e/install-action`, etc.) instead of floating `@v*` tags.
> 
> **Rust CI** gets the meaningful behavior changes: workflow **concurrency** with cancel-in-progress; **mold** for faster links on tests and docs generation; **cargo-tests** and **cargo-docs** move to pinned **`nightly-2026-07-01`** (tests were on stable); **cargo-docs** gains the same **dev** environment, **sccache**, and AWS setup as other Rust jobs. **dtolnay/rust-toolchain** is pinned to a commit with an explicit `toolchain:` input; **`taiki-e/install-action`** installs `just` / `nextest` via `tool:` instead of action aliases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8002641ec1cc59595c61eaca1f05a6d43426a7ce. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by 0xForerunner — APPROVED** [review]

File: .github/workflows/rust-ci.yml [hunk 1]
```diff
@@ -33,6 +33,10 @@ permissions:
   contents: read
   id-token: write
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
```
[the remark this answers is not in the document]
Comment (0xOsiris) [reply .github/workflows/rust-ci.yml:38]: Why would 2 PR's be open for the same branch
